### PR TITLE
Add LegendreGauss points and test

### PR DESCRIPTION
**Pull request by trevor-vincent**
## Proposed changes

1.Generalize CollocationPointsAndWeights class to handle any type of collocation points and weights by templating on the type of collocation point. I had to propagate this change around the code, so many files had to be changed.
2. Add LegendreGauss points as a new type of collocation point.
3. Add UnitTest for LegendreGauss points that tests these points in exactly the same way as LegendreGaussLobatto points.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- [ ] Prefix commits addressing PR requests with `fixup`

### CODE REVIEW

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
Does not follow code review guidelines.

File: src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.hpp [hunk 1]
```diff
@@ -17,12 +17,17 @@ namespace Basis {
 /// \ingroup NumericalAlgorithmsGroup
 /// Functions for using Legendre-Gauss-Lobatto basis
 namespace lgl {
+
+/// enum class used for specifying type of collocation points and weights
+enum class CollocationType { LegendreGauss, LegendreGaussLobatto };
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.hpp]: rename to `Quadrature`

File: src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.hpp [hunk 2]
```diff
@@ -17,12 +17,17 @@ namespace Basis {
 /// \ingroup NumericalAlgorithmsGroup
 /// Functions for using Legendre-Gauss-Lobatto basis
 namespace lgl {
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.hpp]: lgl -> `Legendre` and file name.

File: src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.hpp [hunk 3]
```diff
@@ -17,12 +17,17 @@ namespace Basis {
...
+
+/// enum class used for specifying type of collocation points and weights
+enum class CollocationType { LegendreGauss, LegendreGaussLobatto };
 /// Collocation points.
 /// \param number_of_pts number of collocation points
+template <CollocationType ct>
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.hpp]: Template parameters should be camel case and have useful names

File: src/Domain/LogicalCoordinates.cpp [hunk 4]
```diff
@@ -17,10 +17,13 @@
 template <size_t VolumeDim>
 tnsr::I<DataVector, VolumeDim, Frame::Logical> logical_coordinates(
     const Index<VolumeDim>& extents) {
+  constexpr Basis::lgl::CollocationType LGL =
```
Comment (nilsdeppe) [inline src/Domain/LogicalCoordinates.cpp]: snake_case

File: src/NumericalAlgorithms/LinearOperators/DefiniteIntegral.cpp [hunk 5]
```diff
@@ -14,10 +14,13 @@ namespace {
 template <size_t Dim>
 DataVector integrate_over_last_dimension(const DataVector& integrand,
                                          const Index<Dim>& extents) noexcept {
+  using Basis::lgl::quadrature_weights;
+  constexpr Basis::lgl::CollocationType LGL =
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/LinearOperators/DefiniteIntegral.cpp]: snake_case

File: src/NumericalAlgorithms/LinearOperators/DefiniteIntegral.cpp [hunk 6]
```diff
@@ -14,10 +14,13 @@ namespace {
 template <size_t Dim>
 DataVector integrate_over_last_dimension(const DataVector& integrand,
                                          const Index<Dim>& extents) noexcept {
+  using Basis::lgl::quadrature_weights;
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/LinearOperators/DefiniteIntegral.cpp]: remove

File: src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp [hunk 7]
```diff
@@ -60,6 +60,7 @@ LglQp::LglQp(const size_t poly_degree, const double x) {
   p_ = p_n;
 }
 
+template <CollocationType ct>
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp]: camel case template parameter names

File: tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_LiftFlux.cpp [hunk 8]
```diff
@@ -27,6 +27,10 @@ struct Var {
 
 SPECTRE_TEST_CASE("Unit.DiscontinuousGalerkin.LiftFlux",
                   "[Unit][NumericalAlgorithms]") {
+  using Basis::lgl::quadrature_weights;
```
Comment (nilsdeppe) [inline tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_LiftFlux.cpp]: nope

File: tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_LiftFlux.cpp [hunk 9]
```diff
@@ -27,6 +27,10 @@ struct Var {
...
+  constexpr Basis::lgl::CollocationType LGL =
```
Comment (nilsdeppe) [inline tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_LiftFlux.cpp]: snake_case

File: tests/Unit/NumericalAlgorithms/LinearOperators/Test_DefiniteIntegral.cpp [hunk 10]
```diff
@@ -15,7 +15,10 @@
 namespace {
 void test_definite_integral_1d(const Index<1>& extents) {
   const size_t num_pts_in_x = extents[0];
-  const DataVector& x = Basis::lgl::collocation_points(num_pts_in_x);
+  constexpr Basis::lgl::CollocationType LGL =
```
Comment (nilsdeppe) [inline tests/Unit/NumericalAlgorithms/LinearOperators/Test_DefiniteIntegral.cpp]: snake_case

File: tests/Unit/NumericalAlgorithms/LinearOperators/Test_Linearize.cpp [hunk 11]
```diff
@@ -17,12 +17,15 @@ SPECTRE_TEST_CASE("Unit.Numerical.LinearOperators.Linearize",
                   "[NumericalAlgorithms][LinearOperators][Unit]") {
   // The start and end are chosen to give fast tests
   const size_t n_start = 2, n_end = 5;
+  constexpr Basis::lgl::CollocationType LGL =
```
Comment (nilsdeppe) [inline tests/Unit/NumericalAlgorithms/LinearOperators/Test_Linearize.cpp]: snake_case

File: tests/Unit/NumericalAlgorithms/LinearOperators/Test_MeanValue.cpp [hunk 12]
```diff
@@ -15,12 +15,14 @@
 
 SPECTRE_TEST_CASE("Unit.Numerical.LinearOperators.MeanValue",
                   "[NumericalAlgorithms][LinearOperators][Unit]") {
+  constexpr Basis::lgl::CollocationType LGL =
```
Comment (nilsdeppe) [inline tests/Unit/NumericalAlgorithms/LinearOperators/Test_MeanValue.cpp]: snake_case

File: tests/Unit/NumericalAlgorithms/LinearOperators/Test_PartialDerivatives.cpp [hunk 13]
```diff
@@ -106,8 +106,10 @@ using one_var = typelist<Var1<Dim>>;
 
 template <typename VariableTags, typename GradientTags = VariableTags>
 void test_logical_partial_derivatives_1d(const Index<1>& extents) {
+  constexpr Basis::lgl::CollocationType LGL =
```
Comment (nilsdeppe) [inline tests/Unit/NumericalAlgorithms/LinearOperators/Test_PartialDerivatives.cpp]: snake_case

File: tests/Unit/NumericalAlgorithms/LinearOperators/Test_PartialDerivatives.cpp [hunk 14]
```diff
@@ -133,9 +135,12 @@ void test_logical_partial_derivatives_1d(const Index<1>& extents) {
 
 template <typename VariableTags, typename GradientTags = VariableTags>
 void test_logical_partial_derivatives_2d(const Index<2>& extents) {
+  using CT = Basis::lgl::CollocationType;
```
Comment (nilsdeppe) [inline tests/Unit/NumericalAlgorithms/LinearOperators/Test_PartialDerivatives.cpp]: plz

File: tests/Unit/NumericalAlgorithms/LinearOperators/Test_PartialDerivatives.cpp [hunk 15]
```diff
@@ -168,10 +173,14 @@ void test_logical_partial_derivatives_2d(const Index<2>& extents) {
 
 template <typename VariableTags, typename GradientTags = VariableTags>
 void test_logical_partial_derivatives_3d(const Index<3>& extents) {
+  using CT = Basis::lgl::CollocationType;
```
Comment (nilsdeppe) [inline tests/Unit/NumericalAlgorithms/LinearOperators/Test_PartialDerivatives.cpp]: nerp

File: src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp [hunk 16]
```diff
@@ -80,8 +81,77 @@ class CollocationPointsAndWeights {
...
+
+  DataVector x0(number_of_pts);
+  DataVector a(number_of_pts);
+  DataVector b(number_of_pts);
+  DataVector L0(number_of_pts);
+  DataVector L1(number_of_pts);
+  DataVector L(number_of_pts);
+  DataVector dL(number_of_pts);
+
+  DataVector& x = collocation_pts_;
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp]: really?

File: src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp [hunk 17]
```diff
@@ -80,8 +81,77 @@ class CollocationPointsAndWeights {
...
+template <>
+CollocationPointsAndWeights<CollocationType::LegendreGauss>::
+    CollocationPointsAndWeights(size_t number_of_pts)
+    : collocation_pts_(number_of_pts),
+      quadrature_weights_(number_of_pts),
+      barycentric_weights_(number_of_pts) {
+  // Algorithm 23 from Kopriva's book, p.64
+  double step = 1.;
+  constexpr double tol = 4.0 * std::numeric_limits<double>::epsilon();
+  int iter = 0;
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp]: move this to where it's used.

File: src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp [hunk 18]
```diff
@@ -80,8 +81,77 @@ class CollocationPointsAndWeights {
...
-    const size_t number_of_pts)
+template <>
+CollocationPointsAndWeights<CollocationType::LegendreGauss>::
+    CollocationPointsAndWeights(size_t number_of_pts)
+    : collocation_pts_(number_of_pts),
+      quadrature_weights_(number_of_pts),
+      barycentric_weights_(number_of_pts) {
+  // Algorithm 23 from Kopriva's book, p.64
+  double step = 1.;
+  constexpr double tol = 4.0 * std::numeric_limits<double>::epsilon();
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp]: `constexpr const`

File: src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp [hunk 19]
```diff
@@ -80,8 +81,77 @@ class CollocationPointsAndWeights {
...
-CollocationPointsAndWeights::CollocationPointsAndWeights(
-    const size_t number_of_pts)
+template <>
+CollocationPointsAndWeights<CollocationType::LegendreGauss>::
+    CollocationPointsAndWeights(size_t number_of_pts)
+    : collocation_pts_(number_of_pts),
+      quadrature_weights_(number_of_pts),
+      barycentric_weights_(number_of_pts) {
+  // Algorithm 23 from Kopriva's book, p.64
+  double step = 1.;
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp]: rename to `step_size`

move to where it's used

File: src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp [hunk 20]
```diff
@@ -80,8 +81,77 @@ class CollocationPointsAndWeights {
...
+  DataVector L(number_of_pts);
+  DataVector dL(number_of_pts);
+
+  DataVector& x = collocation_pts_;
+  DataVector& w = quadrature_weights_;
+
+  for (size_t i = 0; i < number_of_pts; ++i) {
+    // Compute Legendre poly recursion coefficients
+    a[i] = (2.0 * i + 1.0) / (i + 1.0);
+    b[i] = static_cast<double>(i) / (i + 1.0);
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp]: remove cast

File: src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp [hunk 21]
```diff
@@ -80,8 +81,77 @@ class CollocationPointsAndWeights {
...
+
+    // set initial guess
+    x[i] = x0[i] = -cos(M_PI * (.75 + i) / (0.5 + number_of_pts));
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp]: 0.75

File: src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp [hunk 22]
```diff
@@ -80,8 +81,77 @@ class CollocationPointsAndWeights {
...
+  }
+
+  while (std::fabs(step) > tol && iter++ < 200) {
+    // Compute Legendre Polynomial and Derivative using recursion
+    for (size_t i = 0; i < number_of_pts; i++) {
+      L0[i] = 1.0;
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp]: use expression templates

File: src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp [hunk 23]
```diff
@@ -80,8 +81,77 @@ class CollocationPointsAndWeights {
...
+      L1[i] = 0.0;
+    }
+
+    for (size_t j = 0; j < number_of_pts; j++) {
+      L1[j] = x[j];
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp]: can L0 and/or L1 be `double`s?

File: src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp [hunk 24]
```diff
@@ -80,8 +81,77 @@ class CollocationPointsAndWeights {
...
+        L[j] = a[k] * x[j] * L1[j] - b[k] * L0[j];
+        L0[j] = L1[j];
+        L1[j] = L[j];
+      }
+      dL[j] = number_of_pts * (L0[j] - x[j] * L[j]) / (1.0 - x[j] * x[j]);
+    }
+
+    // Take a Newton-Raphson step
+    for (size_t i = 0; i < number_of_pts; i++) {
+      step = -L[i] / dL[i];
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp]: "global" `step` -> `max_step`, `step` can be local to here

File: src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp [hunk 25]
```diff
@@ -80,8 +81,77 @@ class CollocationPointsAndWeights {
...
+      x[i] += step;
+      x0[i] = x[i];
+    }
+  }
+
+  ASSERT(std::fabs(step) < tol,
+         "Newton-Raphson failed to reach tolerance for GaussLegendre points");
+
+  for (size_t i = 0; i < number_of_pts; ++i) {
+    double x2 = x[i] * x[i];
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp]: make `const`, fix elsewhere if applicable

File: src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp [hunk 26]
```diff
@@ -80,8 +81,77 @@ class CollocationPointsAndWeights {
...
+    w[i] = 2.0 / ((1 - x2) * (dL[i] * dL[i]));
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp]: remove extra parens

File: src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp [hunk 27]
```diff
@@ -80,8 +81,77 @@ class CollocationPointsAndWeights {
...
+    // explicit barycentric weights from arXiv:1202.0154
+    barycentric_weights_[i] =
+        i % 2 == 0 ? sqrt((1 - x2) * w[i]) : -sqrt((1 - x2) * w[i]);
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp]: 1 -> 1.0

File: src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp (hunk 26 above)
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp]: 1 -> 1.0

File: src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp [hunk 28]
```diff
@@ -80,8 +81,77 @@ class CollocationPointsAndWeights {
...
+  // Algorithm 23 from Kopriva's book, p.64
+  double step = 1.;
+  constexpr double tol = 4.0 * std::numeric_limits<double>::epsilon();
+  int iter = 0;
+
+  DataVector x0(number_of_pts);
+  DataVector a(number_of_pts);
+  DataVector b(number_of_pts);
+  DataVector L0(number_of_pts);
+  DataVector L1(number_of_pts);
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp]: move these down to where they're used

File: src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp [hunk 29]
```diff
@@ -138,13 +208,14 @@ CollocationPointsAndWeights::CollocationPointsAndWeights(
   }
 }
 
-const CollocationPointsAndWeights& collocation_points_and_weights(
+template <CollocationType ct>
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.cpp]: camel case

File: src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.hpp (hunk 2 above)
Comment (nilsdeppe) [reply src/NumericalAlgorithms/Spectral/LegendreGaussLobatto.hpp]: Then class template on `Quadrature` enum

File: tests/Unit/NumericalAlgorithms/Spectral/Test_LegendreGaussLobatto.cpp [hunk 30]
```diff
@@ -35,35 +163,38 @@ SPECTRE_TEST_CASE("Unit.Numerical.Spectral.LegendreGaussLobatto.Points",
        .5349928640318858, .6637764022903113, .7753682609520557,
        .8668779780899503, .9359344988126655, .9807437048939139}};
 
+  constexpr Basis::lgl::CollocationType LGL =
```
Comment (nilsdeppe) [inline tests/Unit/NumericalAlgorithms/Spectral/Test_LegendreGaussLobatto.cpp]: snake_case!!!!

File: tests/Unit/NumericalAlgorithms/Spectral/Test_LegendreGaussLobatto.cpp [hunk 31]
```diff
@@ -12,6 +12,134 @@
 #include "tests/Unit/TestHelpers.hpp"
 
 namespace {
+
+SPECTRE_TEST_CASE("Unit.Numerical.Spectral.LegendreGaussLobatto.GaussPoints",
+                  "[NumericalAlgorithms][Spectral][Unit]") {
+  // Compare GL points to those from
+  // http://people.sc.fsu.edu/~jburkardt/c_src/quadrule/
+
+  constexpr Basis::lgl::CollocationType LG =
```
Comment (nilsdeppe) [inline tests/Unit/NumericalAlgorithms/Spectral/Test_LegendreGaussLobatto.cpp]: Be consistent with GL vs LG on line 18

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

**Comment by trevor-vincent** [comment]
Closing this pull request because we are moving to a different design involving a Legendre class.

**Comment by nilsdeppe** [comment]
This is stale and being superseded by #543

**Pull request closed without merging**